File: pset1/cash/cash.c
#include <math.h>
#include <stdio.h>
#include <stdbool.h>
#include <cs50.h>


int change_calc(float change, int *denomination)
{
    int total = 0;
    int idx = 0;

    while (true)
    {
        if (change < denomination[idx])
        {
            idx++;
            continue;
        }

        int remain = (int)change % denomination[idx];
        int coins = trunc(change / denomination[idx]);

        while (true)
        {
            total++;
            coins--;

            if (coins == 0)
            {
                break;
            }
        }

        if (remain == 0)
        {
            break;
        }

        change = remain;
        idx += 1;
    }
    return total;
}


int main(void)
{
    int denomination[4] = {25, 10, 5, 1};
    float change;

    do
    {
        change = get_float(": ");
    }
    while (change < 0);

    change = (change + .0001) * 100;

    printf("%i\n", change_calc(change, denomination));
}

File: pset6/cash/cash.py
from cs50 import get_float
from itertools import cycle


def change_calc(change, denomination):
    total = 0
    change = int(change * 100)

    for value in cycle(denomination):
    
        if change < value:
            continue

        remain = change % value
        coins = change // value
        
        while True:
            total += 1
            coins -= 1

            if coins == 0:
                break

        if remain == 0:
            break

        change = remain
    return total


def main():
    denomination = [25, 10, 5, 1]

    while True:
    
        change = get_float(": ")

        if change and change > 0:
            break 

    print(change_calc(change, denomination))
    

if __name__ == '__main__':
    main()
